File: 21557 - Mukesh J/Week10/152. Maximum Product Subarray/152. Maximum Product Subarray.java
class Solution {
    public int maxProduct(int[] nums) {
        int minprod= nums[0], maxprod= nums[0], result= nums[0], temp;
        for(int i=1;i<nums.length;i++)
        {
            temp= maxprod;
            maxprod= Math.max(Math.max(maxprod*nums[i], minprod*nums[i]), nums[i]);
            minprod= Math.min(Math.min(temp*nums[i], minprod*nums[i]),nums[i]);
            if(maxprod>result)
                result= maxprod;
        }
        return result;
    }
}

File: 21551-naveen-siva-kumar/week10/152_Maximum Product Subarray.py
class Solution {
public: int maxi(int a,int b,int c){
    if(a>b)
        return a>c?a:c;
    else 
        return b>c?b:c;
}
public: int mini(int a,int b,int c){
    if(a<b)
        return a<c?a:c;
    else 
        return b<c?b:c;
}
public:
    int maxProduct(vector<int>& nums) {
        int max,min,max_till_now,temp;
        max=min=max_till_now=nums[0];
        for(int i=1;i<nums.size();i++){
            temp=max;
            max = maxi(nums[i]*max,nums[i]*min,nums[i]);
            min= mini(nums[i]*temp,nums[i]*min,nums[i]);
            max_till_now =std::max(max_till_now,max);
        }
        return max_till_now;
    }
};
